slateratom/lib/dft.f90

```diff
@@ -279,62 +303,120 @@ subroutine density_grid(pp, max_l, num_alpha, poly_order, alpha, num_mesh_points
 
     ! divide by 4*pi to catch different normalization of spherical harmonics
     rhor(:,:) = transpose(rho) * rec4pi
-    rho4pitot = rho(ii, 1) + rho(ii, 2)
-    if (tGGA .or. tLC) then
+    if (xcFunctional%isGGA(xcnr) .or. xcFunctional%isLongRangeCorrected(xcnr)&
+        & .or. xcFunctional%isGlobalHybrid(xcnr) .or. xcFunctional%isCAMY(xcnr)) then
       sigma(1, :) = drho(:, 1) * drho(:, 1) * rec4pi**2
       sigma(2, :) = drho(:, 1) * drho(:, 2) * rec4pi**2
       sigma(3, :) = drho(:, 2) * drho(:, 2) * rec4pi**2
     end if
 
     select case (xcnr)
-    case(xcFunctional%LDA_PW91)
+    ! LDA-PW91, BNL (long-range corrected)
+    case(xcFunctional%LDA_PW91, xcFunctional%LCY_BNL)
       call xc_f03_lda_exc_vxc(xcfunc_x, nn, rhor(1, 1), ex(1), vx(1, 1))
+    ! GGA-PBE96, GGA-BLYP, LCY-PBE96 (long-range corrected)
     case(xcFunctional%GGA_PBE96, xcFunctional%GGA_BLYP, xcFunctional%LCY_PBE96)
       call xc_f03_gga_exc_vxc(xcfunc_x, nn, rhor(1, 1), sigma(1, 1), ex(1), vx(1,1), vxsigma(1,1))
-      call xc_f03_gga_fxc(xcfunc_x, nn, rhor(1, 1), sigma(1, 1), vx2rho2(1, 1), vx2rhosigma(1, 1),&
-          & vx2sigma2(1, 1))
-    !! exchange of BNL (Slater exchange with Yukawa screening) treated by Vitalijs code
-    case(xcFunctional%LCY_BNL)
-      call xc_f03_lda_exc_vxc(xcfunc_x, nn, rhor(1, 1), ex(1), vx(1, 1))
-      !! important to set to zero, used below for E vs. grad n in GGA
-      vxsigma(:,:) = 0.0_dp
-    case default
-      write(*, '(A,I2,A)') 'XCNR=', xcnr, ' not implemented!'
-      stop
     end select
 
    !! -------- Correlation energy and potential -----------------------
 
     select case (xcnr)
+    ! LDA-PW91
     case(xcFunctional%LDA_PW91)
       call xc_f03_lda_exc_vxc(xcfunc_c, nn, rhor(1, 1), ec(1), vc(1, 1))
       vxc(:,:) = transpose(vx + vc)
+    ! GGA-PBE96, GGA-BLYP, LCY-PBE96, LCY-BNL
     case(xcFunctional%GGA_PBE96, xcFunctional%GGA_BLYP, xcFunctional%LCY_PBE96,&
         & xcFunctional%LCY_BNL)
       call xc_f03_gga_exc_vxc(xcfunc_c, nn, rhor(1, 1), sigma(1, 1), ec(1), vc(1, 1), vcsigma(1, 1))
       vxc(:,:) = transpose(vx + vc)
-      !! derivative of E vs. grad n, treat x and c at the same time
+      ! derivative of E vs. grad n
+      do iSpin = 1, 2
+        ! the other spin
+        iSpin2 = 3 - iSpin
+        ! 1 for spin up, 3 for spin down
+        iSigma = 2 * iSpin - 1
+        tmpv1(:) = (vxsigma(iSigma, :) + vcsigma(iSigma, :)) * drho(:, iSpin) * rec4pi
+        call radial_divergence(tmpv1, abcissa, dz, tmpv2, dzdr)
+        vxc(:, iSpin) = vxc(:, iSpin) - 2.0_dp * tmpv2
+        tmpv1(:) = (vxsigma(2, :) +  vcsigma(2, :)) * drho(:, iSpin2) * rec4pi
+        call radial_divergence(tmpv1, abcissa, dz, tmpv2, dzdr)
+        vxc(:, iSpin) = vxc(:, iSpin) - tmpv2
+      end do
+    end select
+
+    !! -------- Exchange + Correlation energy and potential -----------
+
+    select case (xcnr)
+    ! PBE0, B3LYP (global hybrids), CAMY-B3LYP (CAMY-functional)
+    case(7:9)
```

Comment:
This we should turn into symbolic enumerations!

Reply:
Agree, I must have overlooked these lines back then.